Add optional uncraft time to recipes

pjf opened the pull request:
#### Summary

None

#### Purpose of change

Some recipes, like `seat_leather`, take a long time to create (6 hrs) but certainly shouldn't take that long to disassemble.

#### Describe the solution

Rather than bloating `data/json/uncraft/*` with uncraft recipes with nothing changed but the time, I've added an optional `time` field to the `reversible` JSON field.

You can now do this, to have a seat that takes 6hrs to craft, but only 20 minutes to disassemble:

```
  {
    "type": "recipe",
    "result": "seat_leather",
    "time": "6 h",
    "reversible": { "time": "20 m" },
    ...
  },
```

The existing `"reversible": true` form still works as before.

#### Describe alternatives you've considered

Bloating `data/json/uncraft/*`

#### Testing

- Loaded a game with changes applied.
- Checked recipes in this PR by uncrafting those items
- Checked recipes in this PR still have their correct fabrication times
- Checked other recipes with the default reversible crafting recipes work as before.
- I'm playing with this in my current run.

#### Additional context

Thanks to @moxian for suggesting putting an object inside the reversible field.

Addresses #47199

pjf commented:
Test failure appears unrelated:

```
In update mapgen mx_Trapdoor_spider_den on field, setting terrain to t_region_tree (from t_grass_dead) at (16,19) when furniture was f_dandelion.  Resolve this either by removing the terrain from this mapgen, adding suitable removal commands to the mapgen, or by adding a suitable flag to the innermost mapgen: either ERASE_ALL_BEFORE_PLACING_TERRAIN if you wish terrain to replace everything previously on the tile or ALLOW_TERRAIN_UNDER_OTHER_DATA if you wish the other items to be preserved
0.950 s:      When: the avatar runs the required distance```

moxian commented:
Instead of bloating the recipe with an additional field you can neatly (?) tuck that into the existing `reversible` by making it accept either a boolean or an object with various overrides. So the seat uncraft would look something like:

```json
  {
    "type": "recipe",
    "result": "seat",
    "time": "2 h",
    "reversible": { "time": "20 m" },
    ....
  },
```

This would makes the specification more extensible and allows (in the future, not necessarily in this PR), say, also overriding the tools required. And the components you get out of it.. and other stuff. Potentially even obsoleting the need for `uncraft/` as a whole.

moxian commented on src/recipe.cpp:
```diff
@@ -314,6 +314,21 @@ void recipe::load( const JsonObject &jo, const std::string &src )
+
+            if( jo.has_int( "uncraft_time" ) ) {
+                // Time in moves
+                assign( jo, "uncraft_time", uncraft_time, strict );
```
I think we were moving away from ambiguous integers as time specifiers (are they moves? seconds? or maybe hours?). 
I'd suggest we not introduce new sources of it.

Being able to specify the moves is only *maybe* useful for debugging C++ issues, but at that point you can just override the moves required in C++ proper.

kevingranade replied:
Yea as long as we don't have sub-second uncrafts I'd prefer using time units.

pjf replied:
Excellent! I am delighted for this to be time-unit only!

moxian commented:
(apologies for comment spam) also, please test that recipes with overriden (short) uncrafting times still have proper (long) crafting times.
My C++ is a bit weaker than i'd like, but the way the uncrafts are currently stored make me think they are references to the original ones, rather than the copies. So by changing the craft time of the un- recipe you (maybe?) also change the time of the original. Again, i'd be glad to be proven wrong here.

pjf commented:
> Please test that recipes with overriden (short) uncrafting times still have proper (long) crafting times.

They have their correct times! :tada:

> Instead of bloating the recipe objedcct with an additional field you can neatly (?) tuck that into the existing reversible by making it accept either a boolean or an object with various overrides.

Oh. Oh, I *like* this. Unless anyone tells me there's a good reason not to, I'm 100% running with this suggestion.

pjf commented:
Flipping to draft as I run with @moxian's suggestion to fold the uncraft time data into the `reversible` field, and to remove support for anything but time units.

moxian commented:
thinking more about it, you might be able to leverage the existing machinery that's driving `copy-from` overrides, but i haven't looked at the code so idk, it might be more trouble than it's worth.

pjf commented:
I've updated the code to use the `"reversible": { "time": "20 m" }` form as suggested.

I've retested assembling and disassembling items changed by this PR, as well as ensuring other items with existing `"reversible": "true"` are not affected.

I've removed the ability to have a numeric time for uncrafting. Only time units may be used.

Removing the PR's draft status, as this should be good for review once more. Thanks for the suggestions, everyone!

PatrikLundell commented:
Not really important, but the original post still talks about the original implementation and testing, rather than the modified one. It would be good if it matched the refined implementation and testing of that.

pjf commented:
> Not really important, but the original post still talks about the original implementation and testing, rather than the modified one. It would be good if it matched the refined implementation and testing of that.

Updated as per request!
